src/cider/nrepl/middleware/reload.clj

```diff
@@ -0,0 +1,58 @@
+(ns cider.nrepl.middleware.reload
+  (:require
+   [clj-reload.core :as reload]
+   [clojure.main :refer [repl-caught]]
+   [clojure.string :as str]
+   [haystack.analyzer :as analyzer]
+   [nrepl.middleware.interruptible-eval :refer [*msg*]]
+   [nrepl.middleware.print :as print]
+   [nrepl.misc :refer [response-for]]
+   [nrepl.transport :as transport]
+   [orchard.misc :as misc]))
+
+(defn- user-reload
+  "clj-reload.core/reload from the user project.
+  Must be configured via clj-reload.core/init before being called."
+  []
+  (some-> (symbol "clj-reload.core" "reload") ;; Don't use mrandorsenized version
+          resolve))
+
+(defn- init
+  "Initialize clj-reload with dirs.
+  Only used for test, but necessary because of mranderson."
+  [dirs]
+  (reload/init {:dirs dirs}))
+
+(defn respond
+  [{:keys [transport] :as msg} response]
+  (transport/send transport (response-for msg response)))
+
+(defn- refresh-reply
+  [{:keys [::print/print-fn transport session id] :as msg}]
+  (let [{:keys [exec]} (meta session)]
+    (exec id
+          (fn []
+            (try
+              ;; TODO: clear op
```

Comment:
@tonsky I think I'll wait until you implement the clear op, because I felt my init implementation was a rather far from what you ended up merging and unload seems to be more involved than init.

I think everything else is don't though. Today I tested the e2e of the feature in my emacs and it was working as expected. All users need to do is add `(setq cider-ns-refresh-tool 'clj-reload)`.